src/vulkan-renderer/wrapper/texture.cpp

```diff
@@ -246,4 +260,79 @@ void Texture::create_texture_sampler() {
     spdlog::debug("Image sampler {} created successfully.", m_name);
 }
 
+unsigned char* Texture::load_error_texture_data(int* width, int* height, int* channels)
+{
+    assert(width);
+    assert(height);
+    assert(channels);
+
+    *width = 512;
+    *height = 512;
+    *channels = 4;
+
+    int texture_memory_size = *width * *height * *channels;
+    unsigned char* result = new unsigned char[texture_memory_size];
```

Comment:
I think using an `std::vector` here would be better suited, to avoid the explicit `new` and `delete` calls. You will still be able to use `memcpy`, although I'm not sure if that's the best way.

Reply:
The memcpy comment is very interesting.

So, when I started this, I was doing something like:

```
for y in height
    for x in width
        if ...
```

And, if you want to build a pattern like the following:

```
010
101
010
```
the for-loop if-statement approach works very well.  There is clever modulo logic you can do and it is pretty cool.

The problem I ran into was when you get to something like this:

```
000111000
000111000
000111000
111000111
111000111
111000111
000111000
000111000
000111000
```
I could not figure out a clean and concise for-loop if-statement approach to create that.  I could do it.  But, the ifs became nested with about 4 boolean expressions chained into each if.

So, I don't really know.  That's why I ended up going down the memcpy path.  I couldn't really figure out a nice way to do it with ifs.  (And again, not that the memcpy approach is necessarily simple either.  It is kind of complex too.)

What is also interesting with the memcpy approach, is 90% of the code creates this 25% section of the image:

![image1](https://user-images.githubusercontent.com/4543259/92661225-2669b980-f2ca-11ea-80a1-5584c097a7ec.png)

Then, the last loop finishes the remaining 75%:

![image2](https://user-images.githubusercontent.com/4543259/92661269-41d4c480-f2ca-11ea-8f27-4916775a3d5a.png)

But, even this is not really quite correct.  I don't really need or want 25%.  I would ideally want this:

![image3](https://user-images.githubusercontent.com/4543259/92661525-ce7f8280-f2ca-11ea-9c78-5edb7cd5f531.png)

which is 6.25%.  And then, I could 'stamp' that pattern across the remainder of the texture.

But, it is tricky because that pattern is not contiguous in 1 dimension.  So, I don't know.  This where I sort of stopped aggressively thinking about it and just sort of accepted my current solution.  Unless I am overlooking something obvious and easy, I think it is ironically a kind of tricky problem.

Reply:
I was thinking about the memcpy too, i guess this is easier to understand

```python
def get_color(x: int, y: int, pattern_size: int, colors: int):
    return (int(x / pattern_size) + int(y / pattern_size)) % colors
```
You can use this to determine the color for a pixel.

Reply:
```c++
std::size_t get_color(std::size_t x, std::size_t y, std::size_t square_dimension, std::size_t colors) {
    return (std::size_t(x / square_dimension) + std::size_t(y / square_dimension)) % colors;
}

int main() {
    int width{512};
    int height{512};
    int square_dimension{64};
    
    std::vector<unsigned char> texture{};
    texture.reserve(width * height * 4);

    // pink, purple
    constexpr std::array<std::array<unsigned char, 4>, 2> COLORS{{{0xFF, 0x69, 0xB4, 0xFF}, {0x94, 0x00, 0xD3, 0xFF}}};
    for (std::size_t x = 0; x < width; x++) {
        for (std::size_t y = 0; y < height; y++) {
            const int color_id = get_color(x, y, square_dimension, COLORS.size());
            texture.insert(texture.end(), std::begin(COLORS[color_id]), std::end(COLORS[color_id]));
        }
    }
```
Untested c++ example code.

Reply:
You could use this to create the first two rows and copy them after. (Will reduce the required calculations)

Reply:
Yeah, I think this approach is way better.  This is essentially what I was originally trying to do originally.  However, I could not figure out the `get_color` function linked above.

I will probably remove the memcpy approach entirely.  I think this approach is much more concise and clear.

Reply:
Ignore the first `get_color` python code if you get the c++ code ^^  i was just tired and was about to write the c++ code a day later, but didnt.